### PR TITLE
Copyobject

### PR DESCRIPTION
**Pull request by donatello**
For early review.

### CODE REVIEW

**Comment by harshavardhana** [comment]
@donatello we haven't decided to implement CopyObjectPart API is this a transparent change for CopyObjecr to copy large objects as well ?

**Comment by donatello** [comment]
@harshavardhana Ah, didn't realize copy object part is not implemented in server. Well, as implemented, this api will work for objects upto 5GiB in Minio. It should work for larger objects for S3. I could take up implementing copy-object-part on the server. What say?

**Comment by harshavardhana** [comment]
> @harshavardhana Ah, didn't realize copy object part is not implemented in server. Well, as implemented, this api will work for objects upto 5GiB in Minio. It should work for larger objects for S3. I could take up implementing copy-object-part on the server. What say?

CopyObjectPart is already implemented on server. What i wanted to know is we shouldn't implement a public facing API called as 'CopyObjectPart' its a low level API. 

Are you planning to use CopyObjectPart transparently inside 'CopyObject' for larger objects 5GiB > ?

**Comment by donatello** [comment]
> CopyObjectPart is already implemented on server. What i wanted to know is we shouldn't implement a public facing API called as 'CopyObjectPart' its a low level API.
> 
> Are you planning to use CopyObjectPart transparently inside 'CopyObject' for larger objects 5GiB > ?

Yes, CopyObjectPart is being used transparently inside CopyObject for large objects. So as the server does implement the part copy, it should work transparently to copy objects of any size.

**Comment by harshavardhana** [comment]
> Yes, CopyObjectPart is being used transparently inside CopyObject for large objects. So as the server does implement the part copy, it should work transparently to copy objects of any size.

Great can you open bugs for all libraries to do this as well? @donatello

**Comment by donatello** [comment]
> Great can you open bugs for all libraries to do this as well? @donatello

@harshavardhana will do

**Comment by donatello** [comment]
TODO: pending a test for xmlparsing.

**Comment by donatello** [comment]
> TODO: pending a test for xmlparsing.

This is added now.

File: src/Network/Minio/Data.hs [hunk 1]
```diff
@@ -140,6 +144,47 @@ data ObjectInfo = ObjectInfo {
...
+              fmap formatRFC1123 . cpSourceIfUnmodifiedSince,
+              fmap formatRFC1123 . cpSourceIfModifiedSince]
+    rangeHdr = ("x-amz-copy-source-range",)
+             . HT.renderByteRanges
+             . (:[])
+             . uncurry HT.ByteRangeFromTo
+           <$> (map (both fromIntegral) $
+                maybeToList $ cpSourceRange cps)
+
+-- | Extract the source bucket and source object name. TODO: validate
```
Comment (krisis) [inline src/Network/Minio/Data.hs]: Is this TODO still valid?

**Pull request merged**